### ISSUE TITLE
Add support to XML type

### ISSUE THREAD

**Issue by rafaelreuber** [oracle/python-cx_Oracle#14]
Hi,

The XML type is not supported on cx_Oracle. When I run a code like below, I get the error: AttributeError: 'cx_Oracle.OBJECT' object has no attribute 'read'

```python
result = cursor.execute(
  """SELECT XMLElement("Date", hire_date)
  FROM hr.employees
  WHERE employee_id = 203"""
 )

for res in result:
	# The exception occours right here
	raw_xml = res[0].read()

```

But there are simple workaround. Just use the function **GETCLOBVAL**:

```python
result = cursor.execute(
  """SELECT XMLElement("Date", hire_date).GETCLOBVAL()
  FROM hr.employees
  WHERE employee_id = 203"""
 )

for res in result:
	import xml.etree.ElementTree as ET
	xml_string = res[0].read()
	tree = ET.fromstring(xml_string)
```

The workaround works. But I think that would more pythonic if the cx_Oracle do this automatically:

```python
result = cursor.execute(
  """SELECT XMLElement("Date", hire_date)
  FROM hr.employees
  WHERE employee_id = 203"""
 )

for res in result:
	xml = res[0].read()
	type(xml) # <class 'xml.etree.ElementTree.Element'>
```

**Comment by anthony-tuininga** [comment]
XMLType is returned as an Oracle object with no attributes. It has functions that those can only be called from SQL or PL/SQL -- which defeats the purpose! The workaround you have noted already works for you. Another one is to use an output type handler, as in this code:

```
import cx_Oracle

def OutputTypeHandler(cursor, name, defaultType, size, precision, scale):
    if defaultType == cx_Oracle.OBJECT:
        return cursor.var(cx_Oracle.LONG_STRING, arraysize = cursor.arraysize)

conn = cx_Oracle.Connection("cx_Oracle/dev@localhost/orcl")
conn.outputtypehandler = OutputTypeHandler
cursor = conn.cursor()
cursor.execute("select XMLCol from testxml where rownum <= 1")
row = cursor.fetchone()
print(row)    # will be returned as a string
```

I will look into making this happen automatically since the XMLType objects returned aren't helpful by themselves. In the meantime, hopefully this is a helpful addition.

**Comment by rafaelbezerrac** [comment]
+1

**Comment by zhaowei198311** [comment]
yes

**Comment by gitrevisor** [comment]
Any update on this? Would be really nice to have....

**Comment by anthony-tuininga** [comment]
I did some initial research last year and discussed with the XMLType group internally. A few enhancements to the OCI library are planned which will make working with this type simpler. In the meantime, however, I can look at better workarounds. Thanks for bringing it back to my attention. I'll take another look at it. :-)

**Comment by anthony-tuininga** [comment]
I just created a patch to ODPI-C (which cx_Oracle depends on) which automatically intercepts SYS.XMLTYPE and returns a (LONG) string instead. It is awaiting internal review but I hope to have it available publicly in a few days. Thanks for your patience.

**Comment by gitrevisor** [comment]
Excellent news! Thanks for your hard work!

**Comment by gitrevisor** [comment]
The patch is brilliant except when trying to load xml files larger than 4000 bytes from an Oracle 12c database I get ```ORA-19011: Character string buffer too small```.  Any way around this? Thanks

**Comment by gitrevisor** [comment]
Thanks for your work on this.

**Comment by anthony-tuininga** [comment]
Implemented in cx_Oracle 7.1 which was released yesterday.

**Comment by batatan** [comment]
> The patch is brilliant except when trying to load xml files larger than 4000 bytes from an Oracle 12c database I get `ORA-19011: Character string buffer too small`. Any way around this? Thanks

I have this issue in a django app when I map a  TextField or a CharField to a sys.xmltype field

**Comment by anthony-tuininga** [comment]
You can increase the size to 32,767 bytes if you set [max_string_size](https://docs.oracle.com/database/121/REFRN/GUID-D424D23B-0933-425F-BC69-9C0E6724693C.htm#REFRN10321) to extended. This is available in Oracle Database 12.1 and higher. Unfortunately, there is no other way to increase the size available at the moment. You can, of course, use [getClobVal()](https://docs.oracle.com/en/database/oracle/oracle-database/18/jaxml/oracle/xdb/XMLType.html#getClobVal__) to get the data as a CLOB which is not limited to size -- but isn't seamless either.
